server/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -190,6 +193,9 @@ public IndicesClusterStateService(
         this.client = client;
         this.shardLockRetryInterval = SHARD_LOCK_RETRY_INTERVAL_SETTING.get(settings);
         this.shardLockRetryTimeout = SHARD_LOCK_RETRY_TIMEOUT_SETTING.get(settings);
+
+        // ES-8334 TBD generic is probably wrong for this, we don't want massive parallelism. Maybe MANAGEMENT? Or something throttled?
```

Comment:
I think the main risk here is that the pool is completely depleted by these tasks and thus I'd be in favor of throttling it or a dedicated pool. These threads will basically be just waiting for other work to complete, so a relatively high thread count seems appropriate (10x processors or something like that, unlike MANAGEMENT which has max 5).

I wonder if we'd want to eventually push back on the cluster applier thread if it fills up? Or if we prefer to take that risk and ultimately OOM on it? Hopefully closing will soon be quick(er).

Reply:
It's a little tricky because we might be doing IO on these threads (e.g. flushing) - they're only going to be waiting in the unlucky case where the `IndexWriter` is in the uncancellable step of a merge. So ideally we'd want enough threads to (probably) avoid them all hitting the unlucky case at the same time, without so many that we consume all IO on the node.

My intuition here is that we won't see very many of these tasks in nodes which are running normally, there's not that many shards moving off such nodes, and on a shutting-down node we won't be starting up any new shards so the number of these tasks is bounded by the number of shards to close. The bad case would be a normally-running node with very high churn in shards, starting up new shards so fast that it can't close the old ones down fast enough. Maybe we could block or throttle new shards starting while old shards are still shutting down, given that starting new shards is already async? Still this seems like an unlikely situation, I'd prefer to defer it until later. I'll see if I can track the size of the problem in a metric somehow.

Reply:
Thanks, makes sense, I'd obviously be ok with relatively few threads for this (given that we only have one today) as a starting point.